File: typescript/src/telemetry.ts
/**
 * Optional OpenTelemetry integration.
 *
 * retrykit never imports `@opentelemetry/api` directly, keeping the core free
 * of runtime dependencies and safe to run in the browser, Deno and Bun. Instead
 * it reads the OpenTelemetry API global that the package registers when an SDK
 * is configured. If no provider is registered, every helper here is a no-op.
 */

/** Minimal subset of an OpenTelemetry span that retrykit uses. */
interface MinimalSpan {
  setAttribute(key: string, value: string | number | boolean): unknown;
  recordException(error: unknown): unknown;
  end(): void;
}

/** Minimal subset of an OpenTelemetry tracer that retrykit uses. */
interface MinimalTracer {
  startSpan(name: string): MinimalSpan;
}

interface MinimalTracerProvider {
  getTracer(name: string): MinimalTracer;
}

// OpenTelemetry registers its API singletons on a versioned global symbol.
const OTEL_API_KEY = Symbol.for('opentelemetry.js.api.1');

let overrideTracer: MinimalTracer | null = null;

function lookupTracer(): MinimalTracer | null {
  if (overrideTracer) {
    return overrideTracer;
  }
  const registry = (globalThis as Record<symbol, unknown>)[OTEL_API_KEY] as
    | { trace?: MinimalTracerProvider }
    | undefined;
  const provider = registry?.trace;
  if (provider && typeof provider.getTracer === 'function') {
    return provider.getTracer('retrykit');
  }
  return null;
}

/**
 * Override the tracer used by retrykit.
 *
 * Mainly useful in tests; production code should configure a global
 * OpenTelemetry TracerProvider instead. Pass `null` to clear the override.
 */
export function setTracer(tracer: MinimalTracer | null): void {
  overrideTracer = tracer;
}

/** Whether a tracer is currently available and spans will be created. */
export function isAvailable(): boolean {
  return lookupTracer() !== null;
}

/**
 * Run `fn` inside a span for a single retry attempt.
 *
 * When no tracer is available this calls `fn` directly with no overhead. The
 * span is ended in all cases; a thrown error is recorded before re-throwing.
 *
 * @param name - Span name, typically the wrapped function's name.
 * @param attempt - One-based attempt number, recorded as `retry.attempt`.
 * @param maxAttempts - Total attempt budget, recorded as `retry.max_attempts`.
 * @param fn - The work to execute within the span.
 */
export async function withAttemptSpan<T>(
  name: string,
  attempt: number,
  maxAttempts: number,
  fn: () => Promise<T>,
): Promise<T> {
  const tracer = lookupTracer();
  if (!tracer) {
    return fn();
  }
  const span = tracer.startSpan(`retry.attempt:${name}`);
  span.setAttribute('retry.attempt', attempt);
  span.setAttribute('retry.max_attempts', maxAttempts);
  try {
    const result = await fn();
    span.setAttribute('retry.failed', false);
    return result;
  } catch (error) {
    span.setAttribute('retry.failed', true);
    span.recordException(error);
    throw error;
  } finally {
    span.end();
  }
}

File: python/retrykit/telemetry.py
"""Optional OpenTelemetry integration.

If ``opentelemetry-api`` is installed, retrykit automatically creates a span for
every attempt. If it is not installed, all helpers in this module become no-ops
so that core retrykit has zero required dependencies.
"""

from __future__ import annotations

from collections.abc import Iterator
from contextlib import contextmanager
from typing import Any

try:  # pragma: no cover - exercised indirectly depending on environment
    from opentelemetry import trace as _otel_trace

    _tracer: Any | None = _otel_trace.get_tracer("retrykit")
    OTEL_AVAILABLE = True
except Exception:  # pragma: no cover - import guard
    _otel_trace = None  # type: ignore[assignment]
    _tracer = None
    OTEL_AVAILABLE = False


def is_available() -> bool:
    """Return ``True`` if OpenTelemetry is installed and spans will be created."""
    return OTEL_AVAILABLE


@contextmanager
def attempt_span(
    name: str,
    attempt: int,
    *,
    max_attempts: int | None = None,
) -> Iterator[None]:
    """Create a span for a single retry attempt.

    When OpenTelemetry is unavailable this is a transparent no-op context
    manager, so callers never need to branch on availability themselves.

    Args:
        name: The span name, typically the wrapped callable's qualified name.
        attempt: Zero-based attempt index, recorded as ``retry.attempt``.
        max_attempts: Optional total attempt budget, recorded as
            ``retry.max_attempts``.
    """
    if not OTEL_AVAILABLE or _tracer is None:
        yield
        return

    with _tracer.start_as_current_span(f"retry.attempt:{name}") as span:
        span.set_attribute("retry.attempt", attempt)
        if max_attempts is not None:
            span.set_attribute("retry.max_attempts", max_attempts)
        try:
            yield
        except BaseException as exc:  # noqa: BLE001 - record then re-raise
            span.set_attribute("retry.failed", True)
            span.record_exception(exc)
            raise
        else:
            span.set_attribute("retry.failed", False)


__all__: list[str] = ["OTEL_AVAILABLE", "is_available", "attempt_span"]
